Documentation update about dashes in subnet names

srvvr opened the pull request:
Subnets whose names contain dashes don't get associated to newly created or updated NACLs. Therefore I updated the documentation of this module.

##### SUMMARY
I ran into the behavior that subnets with dashes in their names don't get associated with NACLs. 
Since this seems to be a yaml-behavior that I wasn't aware of, I had to change my subnet names. In the module's documentation dashes are used. My pull request changes and corrects the module documentation accordingly.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Docs Pull Request

##### COMPONENT NAME
~/lib/ansible/modules/cloud/amazon/ec2_vpc_nacl.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
$ ansible --version
ansible 2.4.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/Users/ABC/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Sep 25 2017, 09:53:22) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
```

srvvr committed:
Documentatio update about dashes in subnet names

Subnets whose names contain dashes don't get associated to newly created or updated NACLs. Therefore I updated the documentation of this module.

s-hertel commented:
@srvvr I'm not able to reproduce the issue you experienced with hyphens. Can you provide a reproducer? YAML allows you to use hyphens.

srvvr commented:
Hello Sloane,

Dirk here - in order for you to reproduce this, I can give you the following piece.
I created 3 subnets with a naming convention. For a quicker overview I use just one here.
I created a vars_file that gets used by the creating role. The relevant content is this:

```
vpc:
  name:                     "vpc_{{ organization }}_{{ target_environment }}_{{ zone }}"
  cidr_block:               "{{ cidr_start }}.0.0/16"
  description:              "This vpc hosts the {{ target_environment }} {{ zone }} with all its subnets"
  tags:
    awstype:                "vpc"
    description:            "This VPC hosts the {{ organization|lower }}_{{ target_environment|lower }}_{{ zone|lower }} with all its subnets"
  subnets:
    public:
      cidr_block:           "{{ cidr_start }}.0.0/24"
      region:               "{{ aws_region }}"
      az:                   "{{ aws_az }}"
      tags:
        Name:               "subnet-public-{{ organization }}-{{ target_environment }}-{{ zone }}"
        awstype:            "subnet"
        description:        "This subnet is a publicly reachable subnet. It also hosts a NAT gateway for the other subnets"

.
.
.

nacls:
  - name:                   "nacl_public_{{ organization }}_{{ target_environment }}_{{ zone }}"
    region:                 "{{ aws_region }}"
    tags:
      Name:                 "nacl_public_{{ organization }}_{{ target_environment }}_{{ zone }}"
      awstype:              "nacl"
      description:          "This is the private subnets' network ACL. In this we are as strict as possible to be as safe as possible. Including the restricted SSH access down to the current IP adress"
    subnets:                
      - "{{ vpc.subnets.public.tags.Name }}"
    ingress:
      - [100, 'tcp', 'allow', 'xz.xz.xz.xz/32', null, null, 22, 22]
      - [200, 'icmp', 'allow', 'xz.xz.xz.xz/32', 0, 8]
      - [300, 'tcp', 'allow', '0.0.0.0/0', null, null, 80, 80]
      - [400, 'tcp', 'allow', '0.0.0.0/0', null, null, 443, 443]

```
I use the subnet's name to get an association in a network ACL creating role. The code looks like this:

```
- name:                     "Block for creating network ACLs"
  block:
    - name:                 "Loop for creating network ACLs"
      ec2_vpc_nacl:
        state:              "present"
        vpc_id:             "{{ vpc_facts.vpcs.0.id }}"
        name:               "{{ item.name }}"
        region:             "{{ item.region }}"
        subnets:            "{{ item.subnets }}"
        tags:               "{{ global_tags|combine(item.tags) }}"
        ingress:            "{{ item.ingress }}"
        egress:             "{{ item.egress }}"
      with_items:           "{{ nacls }}"
  when:                     "mode == 'create' and vpc_facts.vpcs != []"

```
When I use the subnet's name with dashes 
`subnet-public-{{ organization }}-{{ target_environment }}-{{ zone }}`

it creates the network ACLs, the subnet associations are missing though.
It is enough to just have one dash in the name. I changed all dashes to underscores and the subnet associations work like a charm. I hope this helps and I hope it is nothing Mac specific. I can live with my workaround, but thought this feedback might help improve the code. I am unfortunately no coder (anymore)...

Thank you and have a great day.
Cheers from Berlin....

 Dirk

dharmabumstead committed:
Small edit.

dharmabumstead reviewed (approved):
Doc update good to go with a small tweak. Thanks @srvvr!

willthames commented:
This seems more like a bug in ec2_vpc_nacl than a documentation bug. I'd rather fix the actual bug.

The code just does the following if the names of subnets are passed in:
```
            subnets = client.describe_subnets(Filters=[
                {'Name': 'tag:Name', 'Values': modules.params['subnets']}])
```

willthames commented:
My test works as expected (tested against v2.4.0.0-1 and devel):
```
TASK [ec2_vpc : debug] **************************************************************************************************************************************************************************
task path: /Users/will/src/ansible-test/ansible/roles/ec2_vpc/tasks/main.yml:20
Tuesday 24 October 2017  19:34:21 +1000 (0:00:03.004)       0:00:04.580 ******* 
ok: [vpc-runner] => {
    "vpc_subnets": {
        "changed": false, 
        "msg": "All items completed", 
        "results": [
            {
                "_ansible_item_result": true, 
                "_ansible_no_log": false, 
                "_ansible_parsed": true, 
                "changed": false, 
                "failed": false, 
                "invocation": {
                    "module_args": {
                        "aws_access_key": null, 
                        "aws_secret_key": null, 
                        "az": "ap-southeast-2a", 
                        "cidr": "10.3.1.0/24", 
                        "ec2_url": null, 
                        "map_public": false, 
                        "profile": "personal", 
                        "region": "ap-southeast-2", 
                        "security_token": null, 
                        "state": "present", 
                        "tags": {
                            "Name": "003-elb-subnet-a", 
                            "Public": "True", 
                            "Tier": "load_balancer"
                        }, 
                        "validate_certs": true, 
                        "vpc_id": "vpc-fe871c9a"
                    }
                }, 
                "item": {
                    "cidr": "10.3.1.0/24", 
                    "tags": {
                        "Name": "003-elb-subnet-a", 
                        "Public": "True", 
                        "Tier": "load_balancer"
                    }, 
                    "zone": "ap-southeast-2a"
                }, 
                "subnet": {
                    "assign_ipv6_address_on_creation": false, 
                    "availability_zone": "ap-southeast-2a", 
                    "available_ip_address_count": 251, 
                    "cidr_block": "10.3.1.0/24", 
                    "default_for_az": false, 
                    "id": "subnet-843b6ef2", 
                    "ipv6_cidr_block_association_set": [], 
                    "map_public_ip_on_launch": false, 
                    "state": "available", 
                    "tags": {
                        "Name": "003-elb-subnet-a", 
                        "Public": "True", 
                        "Tier": "load_balancer"
                    }, 
                    "vpc_id": "vpc-fe871c9a"
                }
            }, 
...
```

```
TASK [ec2_vpc : create NACL] ********************************************************************************************************************************************************************
task path: /Users/will/src/ansible-test/ansible/roles/ec2_vpc/tasks/main.yml:23
Tuesday 24 October 2017  19:34:21 +1000 (0:00:00.052)       0:00:04.632 ******* 
Using module file /Users/will/src/ansible/lib/ansible/modules/cloud/amazon/ec2_vpc_nacl.py
<vpc-runner> ESTABLISH LOCAL CONNECTION FOR USER: will
<vpc-runner> EXEC /bin/sh -c '/usr/local/bin/python && sleep 0'
changed: [vpc-runner] => {
    "changed": true, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "ec2_url": null, 
            "egress": [], 
            "ingress": [], 
            "nacl_id": null, 
            "name": "test-nacl", 
            "profile": "personal", 
            "region": "ap-southeast-2", 
            "security_token": null, 
            "state": "present", 
            "subnets": [
                "003-elb-subnet-a"
            ], 
            "tags": null, 
            "validate_certs": true, 
            "vpc_id": "vpc-fe871c9a"
        }
    }, 
    "nacl_id": "acl-ab9c56cc"
}
```

willthames commented:
If you can show the (possibly sanitised) version of your results using `-vvv`, that would be helpful.

srvvr commented:
If the names simply get passed through to the AWS api via boto (at least this is my limited understanding of the situation), it obviously must be an AWS error. 
Just to clarify - the NACLs are being successfully created and according to the ansible log output, all succeeded. Only within the AWS console (region is eu-central-1 btw) the subnets don't get associated with the new / updated NACLs.
I would still suggest to at least hint to the current situation with AWS in the documentation of the module until the underlying cause has been resolved by Amazon.

srvvr commented:
I created a situation with 2 NACLs and 3 subnets. 
The first NACL (public) gets created but not associated with the subnet that deliberately contains a dash.
The second NACL (private) gets successfully associated with two subnets whose names don't contain dashes. The logfile output with -vvv looks like this
```
TASK [nacl : Loop for creating NACLs] ******************************************************************************************************************************************************
task path: /Users/username/dev/ansible/orgname/aws infrastructure/roles/nacl/tasks/main.yml:14
Using module file /usr/local/lib/python2.7/site-packages/ansible/modules/cloud/amazon/ec2_vpc_nacl.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: username
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/username/.ansible/tmp/ansible-tmp-1508839611.57-199253787595591 `" && echo ansible-tmp-1508839611.57-199253787595591="` echo /Users/username/.ansible/tmp/ansible-tmp-1508839611.57-199253787595591 `" ) && sleep 0'
<localhost> PUT /var/folders/tw/pynzh9d53lb4__xpmyrrgl2xbkvtqy/T/tmpc_bPbO TO /Users/username/.ansible/tmp/ansible-tmp-1508839611.57-199253787595591/ec2_vpc_nacl.py
<localhost> EXEC /bin/sh -c 'chmod u+x /Users/username/.ansible/tmp/ansible-tmp-1508839611.57-199253787595591/ /Users/username/.ansible/tmp/ansible-tmp-1508839611.57-199253787595591/ec2_vpc_nacl.py && sleep 0'
<localhost> EXEC /bin/sh -c 'python /Users/username/.ansible/tmp/ansible-tmp-1508839611.57-199253787595591/ec2_vpc_nacl.py; rm -rf "/Users/username/.ansible/tmp/ansible-tmp-1508839611.57-199253787595591/" > /dev/null 2>&1 && sleep 0'
changed: [localhost] => (item={u'subnets': [u'subnet-public_orgname_dev_intranet'], u'ingress': [[100, u'tcp', u'allow', u'spe.ci.fic.ip/32', None, None, 22, 22], [200, u'icmp', u'allow', u'spe.ci.fic.ip/32', 0, 8], [300, u'tcp', u'allow', u'0.0.0.0/0', None, None, 80, 80], [400, u'tcp', u'allow', u'0.0.0.0/0', None, None, 443, 443]], u'name': u'nacl_public_orgname_dev_intranet', u'tags': {u'awstype': u'nacl', u'Name': u'nacl_public_orgname_dev_intranet', u'description': u"This is the private subnets' network ACL. In this we are as strict as possible to be as safe as possible. Including the restricted SSH access down to the current IP adress"}, u'region': u'eu-central-1', u'egress': [[100, u'tcp', u'allow', u'0.0.0.0/0', None, None, 22, 22], [200, u'tcp', u'allow', u'0.0.0.0/0', None, None, 80, 80], [300, u'tcp', u'allow', u'0.0.0.0/0', None, None, 443, 443], [400, u'udp', u'allow', u'0.0.0.0/0', None, None, 53, 53], [500, u'udp', u'allow', u'0.0.0.0/0', None, None, 123, 123]]}) => {
    "changed": true, 
    "failed": false, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "ec2_url": null, 
            "egress": [
                [
                    100, 
                    "tcp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    22, 
                    22
                ], 
                [
                    200, 
                    "tcp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    80, 
                    80
                ], 
                [
                    300, 
                    "tcp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    443, 
                    443
                ], 
                [
                    400, 
                    "udp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    53, 
                    53
                ], 
                [
                    500, 
                    "udp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    123, 
                    123
                ]
            ], 
            "ingress": [
                [
                    100, 
                    "tcp", 
                    "allow", 
                    "spe.ci.fic.ip/32", 
                    null, 
                    null, 
                    22, 
                    22
                ], 
                [
                    200, 
                    "icmp", 
                    "allow", 
                    "spe.ci.fic.ip/32", 
                    0, 
                    8
                ], 
                [
                    300, 
                    "tcp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    80, 
                    80
                ], 
                [
                    400, 
                    "tcp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    443, 
                    443
                ]
            ], 
            "nacl_id": null, 
            "name": "nacl_public_orgname_dev_intranet", 
            "profile": null, 
            "region": "eu-central-1", 
            "security_token": null, 
            "state": "present", 
            "subnets": [
                "subnet-public_orgname_dev_intranet"
            ], 
            "tags": {
                "Name": "nacl_public_orgname_dev_intranet", 
                "awstype": "nacl", 
                "description": "This is the private subnets' network ACL. In this we are as strict as possible to be as safe as possible. Including the restricted SSH access down to the current IP adress", 
                "environment": "dev", 
                "organization": "orgname", 
                "zone": "intranet"
            }, 
            "validate_certs": true, 
            "vpc_id": "vpc-fc424c94"
        }
    }, 
    "item": {
        "egress": [
            [
                100, 
                "tcp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                22, 
                22
            ], 
            [
                200, 
                "tcp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                80, 
                80
            ], 
            [
                300, 
                "tcp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                443, 
                443
            ], 
            [
                400, 
                "udp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                53, 
                53
            ], 
            [
                500, 
                "udp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                123, 
                123
            ]
        ], 
        "ingress": [
            [
                100, 
                "tcp", 
                "allow", 
                "spe.ci.fic.ip/32", 
                null, 
                null, 
                22, 
                22
            ], 
            [
                200, 
                "icmp", 
                "allow", 
                "spe.ci.fic.ip/32", 
                0, 
                8
            ], 
            [
                300, 
                "tcp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                80, 
                80
            ], 
            [
                400, 
                "tcp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                443, 
                443
            ]
        ], 
        "name": "nacl_public_orgname_dev_intranet", 
        "region": "eu-central-1", 
        "subnets": [
            "subnet-public_orgname_dev_intranet"
        ], 
        "tags": {
            "Name": "nacl_public_orgname_dev_intranet", 
            "awstype": "nacl", 
            "description": "This is the private subnets' network ACL. In this we are as strict as possible to be as safe as possible. Including the restricted SSH access down to the current IP adress"
        }
    }, 
    "nacl_id": "acl-e99c5182"
}
Using module file /usr/local/lib/python2.7/site-packages/ansible/modules/cloud/amazon/ec2_vpc_nacl.py
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/username/.ansible/tmp/ansible-tmp-1508839613.86-1912083550857 `" && echo ansible-tmp-1508839613.86-1912083550857="` echo /Users/username/.ansible/tmp/ansible-tmp-1508839613.86-1912083550857 `" ) && sleep 0'
<localhost> PUT /var/folders/tw/pynzh9d53lb4__xpmyrrgl2xbkvtqy/T/tmplySZbn TO /Users/username/.ansible/tmp/ansible-tmp-1508839613.86-1912083550857/ec2_vpc_nacl.py
<localhost> EXEC /bin/sh -c 'chmod u+x /Users/username/.ansible/tmp/ansible-tmp-1508839613.86-1912083550857/ /Users/username/.ansible/tmp/ansible-tmp-1508839613.86-1912083550857/ec2_vpc_nacl.py && sleep 0'
<localhost> EXEC /bin/sh -c 'python /Users/username/.ansible/tmp/ansible-tmp-1508839613.86-1912083550857/ec2_vpc_nacl.py; rm -rf "/Users/username/.ansible/tmp/ansible-tmp-1508839613.86-1912083550857/" > /dev/null 2>&1 && sleep 0'
changed: [localhost] => (item={u'subnets': [u'subnet_app_orgname_dev_intranet', u'subnet_data_orgname_dev_intranet'], u'ingress': [[100, u'tcp', u'allow', u'77.77.0.0/16', None, None, 22, 22], [200, u'icmp', u'allow', u'77.77.0.0/16', 0, 8], [300, u'tcp', u'allow', u'77.77.0.0/16', None, None, 443, 443]], u'name': u'nacl_private_orgname_dev_intranet', u'tags': {u'awstype': u'nacl', u'Name': u'nacl_private_orgname_dev_intranet', u'description': u"This is the privte subnets' network ACL. "}, u'region': u'eu-central-1', u'egress': [[100, u'tcp', u'allow', u'0.0.0.0/0', None, None, 22, 22], [200, u'tcp', u'allow', u'0.0.0.0/0', None, None, 80, 80], [300, u'tcp', u'allow', u'0.0.0.0/0', None, None, 443, 443], [400, u'udp', u'allow', u'0.0.0.0/0', None, None, 53, 53], [500, u'udp', u'allow', u'0.0.0.0/0', None, None, 123, 123]]}) => {
    "changed": true, 
    "failed": false, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "ec2_url": null, 
            "egress": [
                [
                    100, 
                    "tcp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    22, 
                    22
                ], 
                [
                    200, 
                    "tcp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    80, 
                    80
                ], 
                [
                    300, 
                    "tcp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    443, 
                    443
                ], 
                [
                    400, 
                    "udp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    53, 
                    53
                ], 
                [
                    500, 
                    "udp", 
                    "allow", 
                    "0.0.0.0/0", 
                    null, 
                    null, 
                    123, 
                    123
                ]
            ], 
            "ingress": [
                [
                    100, 
                    "tcp", 
                    "allow", 
                    "77.77.0.0/16", 
                    null, 
                    null, 
                    22, 
                    22
                ], 
                [
                    200, 
                    "icmp", 
                    "allow", 
                    "77.77.0.0/16", 
                    0, 
                    8
                ], 
                [
                    300, 
                    "tcp", 
                    "allow", 
                    "77.77.0.0/16", 
                    null, 
                    null, 
                    443, 
                    443
                ]
            ], 
            "nacl_id": null, 
            "name": "nacl_private_orgname_dev_intranet", 
            "profile": null, 
            "region": "eu-central-1", 
            "security_token": null, 
            "state": "present", 
            "subnets": [
                "subnet_app_orgname_dev_intranet", 
                "subnet_data_orgname_dev_intranet"
            ], 
            "tags": {
                "Name": "nacl_private_orgname_dev_intranet", 
                "awstype": "nacl", 
                "description": "This is the privte subnets' network ACL. ", 
                "environment": "dev", 
                "organization": "orgname", 
                "zone": "intranet"
            }, 
            "validate_certs": true, 
            "vpc_id": "vpc-fc424c94"
        }
    }, 
    "item": {
        "egress": [
            [
                100, 
                "tcp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                22, 
                22
            ], 
            [
                200, 
                "tcp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                80, 
                80
            ], 
            [
                300, 
                "tcp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                443, 
                443
            ], 
            [
                400, 
                "udp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                53, 
                53
            ], 
            [
                500, 
                "udp", 
                "allow", 
                "0.0.0.0/0", 
                null, 
                null, 
                123, 
                123
            ]
        ], 
        "ingress": [
            [
                100, 
                "tcp", 
                "allow", 
                "77.77.0.0/16", 
                null, 
                null, 
                22, 
                22
            ], 
            [
                200, 
                "icmp", 
                "allow", 
                "77.77.0.0/16", 
                0, 
                8
            ], 
            [
                300, 
                "tcp", 
                "allow", 
                "77.77.0.0/16", 
                null, 
                null, 
                443, 
                443
            ]
        ], 
        "name": "nacl_private_orgname_dev_intranet", 
        "region": "eu-central-1", 
        "subnets": [
            "subnet_app_orgname_dev_intranet", 
            "subnet_data_orgname_dev_intranet"
        ], 
        "tags": {
            "Name": "nacl_private_orgname_dev_intranet", 
            "awstype": "nacl", 
            "description": "This is the privte subnets' network ACL. "
        }
    }, 
    "nacl_id": "acl-00824f6b"
}
```

The result in the AWS console looks like this:
![image](https://user-images.githubusercontent.com/32300982/31938244-4dc62298-b8b7-11e7-9216-55f82a44ff65.png)

willthames commented:
I edited your comment to use triple backticks rather than single backticks so that the output is actually readable :)

srvvr commented:
Thank you for correcting the log output and sorry for the unreadable version.

willthames commented:
No problem. 

I just can't reproduce this. My NACLs with dashes are showing up in the console exactly as expected, associated with the correct subnets. 

Immediately after running ec2_vpc_nacl, it might be worth doing

```
ec2_vpc_nacl_facts:
  filters:
    tag:Name: "nacl_private_orgname_dev_intranet"
    tag:environment: dev
  region: "eu-central-1"
```

or `aws ec2 describe-network-acls --region eu-central-1 --filters Name=tag:Name,Values=nacl_private_orgname_dev_intranet`

If the results of both of those are as expected, then it would seem it's a presentation issue on the AWS side.

srvvrs commented:
The NACL that should be associated with the subnet with dashes in its name comes back empty when I list it with ec2_vpc_nacl_facts
```
...
"subnets": [],
...

```
Same with 

```aws ec2 describe-network-acls --region eu-central-1```

the subnet with the dash in its name is still associated with the default subnet.
```
        {
            "Associations": [
                {
                    "SubnetId": "subnet-a82e3cd2", 
                    "NetworkAclId": "acl-45c4082e", 
                    "NetworkAclAssociationId": "aclassoc-a99394c4"
                }
            ], 
            "NetworkAclId": "acl-45c4082e", 
            "VpcId": "vpc-ea6f6682", 
            "Tags": [], 
            "Entries": [
                {
                    "RuleNumber": 100, 
                    "Protocol": "-1", 
                    "Egress": true, 
                    "CidrBlock": "0.0.0.0/0", 
                    "RuleAction": "allow"
                }, 
                {
                    "RuleNumber": 32767, 
                    "Protocol": "-1", 
                    "Egress": true, 
                    "CidrBlock": "0.0.0.0/0", 
                    "RuleAction": "deny"
                }, 
                {
                    "RuleNumber": 100, 
                    "Protocol": "-1", 
                    "Egress": false, 
                    "CidrBlock": "0.0.0.0/0", 
                    "RuleAction": "allow"
                }, 
                {
                    "RuleNumber": 32767, 
                    "Protocol": "-1", 
                    "Egress": false, 
                    "CidrBlock": "0.0.0.0/0", 
                    "RuleAction": "deny"
                }
            ], 
            "IsDefault": true
        }
```
The designated NACL doesn't have any associations :/

```
{
    "NetworkAcls": [
        {
            "Associations": [], 
            "NetworkAclId": "acl-82c408e9", 
            "VpcId": "vpc-ea6f6682", 
            "Tags": [
                {
                    "Value": "This is the public subnets' network ACL. In this we are as strict as possible to be as safe as possible. Including the restricted SSH access down to the current IP adress", 
                    "Key": "description"
                }, ...
```

This is the same result as displayed in the AWS console (see screenshot).
I assume it could be an error with AWS in the eu-central-1 region?!

willthames commented:
Ok, this is progress. You are getting the results back from the CLI and ec2_vpc_nacl_facts that match the console. You are hitting some kind of side-effect that I am not hitting.

Can you try running the following with ansible-playbook playbook.yml -vvv (replacing VPC_ID and TEST_SUBNET (this subnet should not be an important one as the subnet should get the test NACL associated with it) appropriately)

```
- hosts: localhost
  connection: local

  tasks:
  - ec2_vpc_nacl:
      region: eu-central-1
      name: test-nacl-123
      vpc_id: vpc-VPC_ID
      subnets: TEST_SUBNET
    register: nacl_result

  - ec2_vpc_nacl_facts:
      region: eu-central-1
      nacl_ids: "{{ nacl_result.nacl_id }}"
```

and just to be on the safe side

```
ls -l /Users/ABC/.ansible/plugins/modules /usr/share/ansible/plugins/modules
```

Running the above playbook worked as I expected (even when I also used your ingress/egress rules)

srvvrs commented:
Same, same...
when I run the playbook with the vpc-id and the subnet's name, I get the following output:

```
$ ansible-playbook playbook.yml -vvv
ansible-playbook 2.4.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/Users/ABC/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible-playbook
  python version = 2.7.14 (default, Sep 25 2017, 09:53:22) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
Using /etc/ansible/ansible.cfg as config file
 [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: Could not match supplied host pattern, ignoring: all

 [WARNING]: provided hosts list is empty, only localhost is available


PLAYBOOK: playbook.yml *************************************************************************************************************************
1 plays in playbook.yml

PLAY [localhost] *******************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************
Using module file /usr/local/lib/python2.7/site-packages/ansible/modules/system/setup.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: ABC
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/ABC/.ansible/tmp/ansible-tmp-1508932088.57-20618927658975 `" && echo ansible-tmp-1508932088.57-20618927658975="` echo /Users/ABC/.ansible/tmp/ansible-tmp-1508932088.57-20618927658975 `" ) && sleep 0'
<127.0.0.1> PUT /var/folders/tw/pynzh9d53lb4__xpmyrrgl2xbkvtqy/T/tmpsD0z6v TO /Users/ABC/.ansible/tmp/ansible-tmp-1508932088.57-20618927658975/setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/ABC/.ansible/tmp/ansible-tmp-1508932088.57-20618927658975/ /Users/ABC/.ansible/tmp/ansible-tmp-1508932088.57-20618927658975/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/local/opt/python/bin/python2.7 /Users/ABC/.ansible/tmp/ansible-tmp-1508932088.57-20618927658975/setup.py; rm -rf "/Users/ABC/.ansible/tmp/ansible-tmp-1508932088.57-20618927658975/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]
META: ran handlers

TASK [ec2_vpc_nacl] ****************************************************************************************************************************
task path: /Users/ABC/dev/ansible/organization/aws infrastructure/playbook.yml:5
Using module file /usr/local/lib/python2.7/site-packages/ansible/modules/cloud/amazon/ec2_vpc_nacl.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: ABC
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/ABC/.ansible/tmp/ansible-tmp-1508932089.3-73443927395615 `" && echo ansible-tmp-1508932089.3-73443927395615="` echo /Users/ABC/.ansible/tmp/ansible-tmp-1508932089.3-73443927395615 `" ) && sleep 0'
<127.0.0.1> PUT /var/folders/tw/pynzh9d53lb4__xpmyrrgl2xbkvtqy/T/tmpaxMJCk TO /Users/ABC/.ansible/tmp/ansible-tmp-1508932089.3-73443927395615/ec2_vpc_nacl.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/ABC/.ansible/tmp/ansible-tmp-1508932089.3-73443927395615/ /Users/ABC/.ansible/tmp/ansible-tmp-1508932089.3-73443927395615/ec2_vpc_nacl.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/local/opt/python/bin/python2.7 /Users/ABC/.ansible/tmp/ansible-tmp-1508932089.3-73443927395615/ec2_vpc_nacl.py; rm -rf "/Users/ABC/.ansible/tmp/ansible-tmp-1508932089.3-73443927395615/" > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true,
    "failed": false,
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "ec2_url": null,
            "egress": [],
            "ingress": [],
            "nacl_id": null,
            "name": "test-nacl-123",
            "profile": null,
            "region": "eu-central-1",
            "security_token": null,
            "state": "present",
            "subnets": [
                "subnet-public_organization_dev_intranet"
            ],
            "tags": null,
            "validate_certs": true,
            "vpc_id": "vpc-106e6778"
        }
    },
    "nacl_id": "acl-30da165b"
}

TASK [ec2_vpc_nacl_facts] **********************************************************************************************************************
task path: /Users/ABC/dev/ansible/organization/aws infrastructure/playbook.yml:12
Using module file /usr/local/lib/python2.7/site-packages/ansible/modules/cloud/amazon/ec2_vpc_nacl_facts.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: ABC
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/ABC/.ansible/tmp/ansible-tmp-1508932091.95-80938868093826 `" && echo ansible-tmp-1508932091.95-80938868093826="` echo /Users/ABC/.ansible/tmp/ansible-tmp-1508932091.95-80938868093826 `" ) && sleep 0'
<127.0.0.1> PUT /var/folders/tw/pynzh9d53lb4__xpmyrrgl2xbkvtqy/T/tmp7RAqJN TO /Users/ABC/.ansible/tmp/ansible-tmp-1508932091.95-80938868093826/ec2_vpc_nacl_facts.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/ABC/.ansible/tmp/ansible-tmp-1508932091.95-80938868093826/ /Users/ABC/.ansible/tmp/ansible-tmp-1508932091.95-80938868093826/ec2_vpc_nacl_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/local/opt/python/bin/python2.7 /Users/ABC/.ansible/tmp/ansible-tmp-1508932091.95-80938868093826/ec2_vpc_nacl_facts.py; rm -rf "/Users/ABC/.ansible/tmp/ansible-tmp-1508932091.95-80938868093826/" > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "changed": false,
    "failed": false,
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "ec2_url": null,
            "filters": {},
            "nacl_ids": [
                "acl-30da165b"
            ],
            "profile": null,
            "region": "eu-central-1",
            "security_token": null,
            "validate_certs": true
        }
    },
    "nacls": [
        {
            "egress": [],
            "ingress": [],
            "is_default": false,
            "nacl_id": "acl-30da165b",
            "subnets": [],
            "tags": {
                "Name": "test-nacl-123"
            },
            "vpc_id": "vpc-106e6778"
        }
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0
```

the ls-command results in no relevant output:
```
$ ls -l /Users/ABC/.ansible/plugins/modules /usr/share/ansible/plugins/modules
ls: /Users/ABC/.ansible/plugins/modules: No such file or directory
ls: /usr/share/ansible/plugins/modules: No such file or directory
```
 All ansible modules are actually located under 
`/usr/local/lib/python2.7/site-packages/ansible/modules/`

I hope this helps....

willthames commented:
> I hope this helps....

Haha, it massively adds to the mystery! But it's all useful information. Even if I'm completely none the wiser right now.

Can you do

```
aws ec2 describe-subnet --region eu-central-1 --subnet-id subnet-a82e3cd2
```

assuming that is the subnet ID for the subnet with the dash.

srvvrs commented:
This results in the following:
```
$ aws ec2 describe-subnets --region eu-central-1 --subnet-id subnet-f5d2c08f
{
    "Subnets": [
        {
            "AvailabilityZone": "eu-central-1b",
            "Tags": [
                {
                    "Value": "subnet-public_organization_dev_intranet",
                    "Key": "Name"
                },
                {
                    "Value": "dev",
                    "Key": "environment"
                },
                {
                    "Value": "intranet",
                    "Key": "zone"
                },
                {
                    "Value": "subnet",
                    "Key": "awstype"
                },
                {
                    "Value": "organization",
                    "Key": "organization"
                },
                {
                    "Value": "This subnet is a publicly reachable subnet. It also hosts a NAT gateway for the other subnets",
                    "Key": "description"
                }
            ],
            "AvailableIpAddressCount": 250,
            "DefaultForAz": false,
            "Ipv6CidrBlockAssociationSet": [],
            "VpcId": "vpc-106e6778",
            "State": "available",
            "MapPublicIpOnLaunch": false,
            "SubnetId": "subnet-f5d2c08f",
            "CidrBlock": "77.77.0.0/24",
            "AssignIpv6AddressOnCreation": false
        }
    ]
}
```

srvvrs commented:
In case you should wonder why the IDs are changing - I am writing infrastructure plays/roles that I run every now and then. So the IDs are a moving target.....

willthames commented:
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_vpc_nacl.py#L523

Oh, it's because it starts with `subnet-`. It thinks it's a subnet ID. We can consider this a bug (it wouldn't be hard to do `subnet-[a-f0-9]{8}` or similar) but not a docs bug. Could you raise a new issue for this and I'll get it fixed.

srvvrs commented:
Hahahahaha - ok - that was a stupid naming approach on my end too. So I'll just rename the first part. I wouldn't bother with a regexp. I bet if that was in there, I would have still magically hit that one with an even stupider naming pattern. 
Plus if AWS at one point extends their IDs, this will crash. So we'd create a certain future error. I would still just point out, that a subnet's name shouldn't start with `/^subnet-/`
Thank you so much for your investigation :D

willthames commented:
If you want to update this PR to fix the docs to say that, I can reopen it.

srvvrs commented:
I actually opened a new issue
https://github.com/ansible/ansible/issues/32133#issuecomment-339315224 <https://github.com/ansible/ansible/issues/32133#issuecomment-339315224>
suggesting to check for subnet-name first, which will always work.

I will update the PR too...

> On 25 Oct 2017, at 14:51, Will Thames <notifications@github.com> wrote:
> 
> If you want to update this PR to fix the docs to say that, I can reopen it.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/ansible/ansible/pull/32034#issuecomment-339319268>, or mute the thread <https://github.com/notifications/unsubscribe-auth/Aezd7OPV6McV8nzH-I3dV1Nmpg_JthS8ks5svy7SgaJpZM4QCqYm>.
>
